rules/CodeQuality/Rector/ClassConstFetch/ConvertStaticPrivateConstantToSelfRector.php

```diff
@@ -0,0 +1,122 @@
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @see \Rector\Tests\CodeQuality\Rector\ClassConstFetch\ConvertStaticPrivateConstantToSelfRector\ConvertStaticPrivateConstantToSelfRectorTest
+ * @see https://3v4l.org/8Y0ba
+ * @see https://phpstan.org/r/11d4c850-1a40-4fae-b665-291f96104d11
+ */
+final class ConvertStaticPrivateConstantToSelfRector extends AbstractRector
+{
+    public function __construct(
+        private readonly ReflectionResolver $reflectionResolver
+    ) {
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Replaces static::* access to private constants with self::* on final classes',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+final class Foo {
+    private const BAR = 'bar';
+    public function run()
+    {
+        $bar = static::BAR;
+    }
+}
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+final class Foo {
+    private const BAR = 'bar';
+    public function run()
+    {
+        $bar = self::BAR;
+    }
+}
+CODE_SAMPLE
+                    ,
+                ),
+            ],
+        );
+    }
+
+    public function getNodeTypes(): array
+    {
+        return [ClassConstFetch::class];
+    }
+
+    /**
+     * @param \PhpParser\Node\Expr\ClassConstFetch $node
+     */
+    public function refactor(Node $node): ?ClassConstFetch
+    {
+        if (! $this->isUsingStatic($node)) {
+            return null;
+        }
+
+        if (! $this->isPrivateConstant($node)) {
+            return null;
+        }
+
+        return new ClassConstFetch(new Name('self'), $node->name);
+    }
+
+    private function isUsingStatic(ClassConstFetch $node): bool
+    {
+        if (! $node->class instanceof Name) {
+            return false;
+        }
+
+        return $node->class->getFirst() === 'static';
+    }
+
+    private function isPrivateConstant(ClassConstFetch $node): bool
+    {
+        $scope = $node->getAttribute(AttributeKey::SCOPE);
+        if (! $scope instanceof Scope) {
+            return false;
+        }
+        $classReflection = $this->reflectionResolver->resolveClassReflection($node);
+        if (! $classReflection instanceof ClassReflection) {
+            return false;
+        }
+        $class = $this->betterNodeFinder->findParentType($node, Node\Stmt\Class_::class);
+        if (! $class instanceof Node\Stmt\Class_) {
+            return false;
+        }
+        if (! $classReflection->isClass() || ! $class->isFinal()) {
+            return false;
+        }
+        $constantName = $node->name;
+        if ($constantName instanceof Error) {
+            return false;
+        }
+        try {
+            $constantReflection = $classReflection->getConstant($constantName->toString());
+        } catch (Exception $e) {
+            return false;
+        }
```

Comment:
There is `ClassInsertManipulator->hasClassConstant(Class_ $class, string $constantName)` for it.

Reply:
I see, the constant need to be get, in that case, loop and get can be used:

```php
        $name = $constantName->toString();
        foreach ($class->getConstants() as $classConst) {
            if (! $this->nodeNameResolver->isName($classConst, $name)) {
                return true;
            }
           
           return $classConst->isPrivate();
        }

       return false;
```

Reply:
You probably meant this, right?
```diff
        $name = $constantName->toString();
        foreach ($class->getConstants() as $classConst) {
            if (! $this->nodeNameResolver->isName($classConst, $name)) {
-                return true;
+                return false;
            }
           
           return $classConst->isPrivate();
        }

       return false;
```

Reply:
Done